runtime/common/src/apis.rs

```diff
@@ -351,20 +351,29 @@ macro_rules! impl_runtime_apis_plus_common {
 					slot: u32,
 					parent_header: &<Block as BlockT>::Header
 				) -> bool {
+					let block_number = parent_header.number + 1;
+
 					// The Moonbeam runtimes use an entropy source that needs to do some accounting
 					// work during block initialization. Therefore we initialize it here to match
 					// the state it will be in when the next block is being executed.
 					use frame_support::traits::OnInitialize;
 					System::initialize(
-						&(parent_header.number + 1),
+						&block_number,
 						&parent_header.hash(),
 						&parent_header.digest,
 						frame_system::InitKind::Inspection
 					);
-					RandomnessCollectiveFlip::on_initialize(System::block_number());
-
-					// And now the actual prediction call
-					AuthorInherent::can_author(&author, &slot)
+					RandomnessCollectiveFlip::on_initialize(block_number);
+
+					// Because the staking solution calculates the next staking set at the beginning
+					// of the first block in the new round, the only way to accurately predict the
+					// authors would be to run the staking election while predicting. However this
+					// election is heavy and will take too long during prediction. So instead we
+					// work around it by always authoring the first slot in a new round. A longer-
+					// term solution will be to calculate the staking election result in the last
+					// block of the ending round.
+					parachain_staking::Pallet::<Self>::round().should_update(block_number)
+					|| AuthorInherent::can_author(&author, &slot)
```

Comment:
It might be more efficient to swap these since `should_update()` will return false more often than `can_author()`

Reply:
Good suggestion, made me wonder if rust short-circuits `||` statements and it does, I tried it https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=cbfb167666248e27f10d5394519cb080

Reply:
That code you pasted is actually a pretty common pattern that I've seen before...

Reply:
I like this idea, but I couldn't figure out how to do it, and also reliably print the warning message that you recommended, and I decided that this was not that significant of an optimization, so I opted for the message instead. Feel free to suggest a way to have both.

_sits back and waits for the Rust ninjas_

Reply:
I'll take a stab at it. Mind if I push to your branch?